src/providers/SamlAuthProvider.ts

```diff
@@ -0,0 +1,291 @@
+   */
+  createSamlRedirectUrl(stateId: UniqueID): URL {
+    const publisher = this._context.extension.packageJSON.publisher;
+    const name = this._context.extension.packageJSON.name;
+    return new URL(`${vscode.env.uriScheme}://${publisher}.${name}/${stateId}`);
+  }
+
+  dispose = (): void => {
+    this._disposable.dispose();
+  };
+
+  /**
+   * Get a list of sessions.
+   * @param scopes An optional list of scopes. If provided, the sessions returned
+   * these scopes, otherwise all sessions should be returned.
+   * @returns A promise that resolves to an array of authentication sessions.
+   */
+  getSessions = async (
+    _scopes?: readonly string[]
+  ): Promise<readonly vscode.AuthenticationSession[]> => {
+    // Not implementing this since there really isn't a concept of persistent
+    // SAML sessions that we can store in the extension. Once the user disconnects,
+    // there's no way to re-use the session since the extension only has access
+    // to the short lived nonce used to login. IdPs often store session info in
+    // browser, so user can just re-login, and if they have an active session,
+    // the browser will just pass it along without requiring credentials again.
+    return [];
+  };
+
+  /**
+   * Prompts a user to login.
+   *
+   * If login is successful, the onDidChangeSessions event should be fired.
+   *
+   * If login fails, a rejected promise should be returned.
+   *
+   * If the provider has specified that it does not support multiple accounts,
+   * then this should never be called if there is already an existing session
+   * matching these scopes.
+   * @param scopes A list of scopes that the new session should be created with.
+   * @returns A promise that resolves to an authentication session.
+   */
+  createSession = async (
+    scopes: readonly string[]
+  ): Promise<vscode.AuthenticationSession> => {
+    const samlScopes = parseSamlScopes(scopes);
+    if (samlScopes == null) {
+      throw new Error(
+        'SAML authentication provider does not support this scope.'
+      );
+    }
+
+    logger.debug('createSession', samlScopes);
+
+    const samlSessionKey = makeSAMLSessionKey();
+    logger.debug('samlSessionKey:', `${samlSessionKey}`);
+
+    const serverUrl = new URL(samlScopes.serverUrl);
+    const { stateId } = SamlAuthProvider.pendingAuthState.getOrThrow(serverUrl);
+
+    const samlLoginUrl = new URL(samlScopes.samlLoginUrl);
+    samlLoginUrl.searchParams.append('key', samlSessionKey);
+    samlLoginUrl.searchParams.append(
+      'redirect',
+      `${this.createSamlRedirectUrl(stateId)}`
+    );
+
+    const authSucceeded = await vscode.window.withProgress<boolean>(
+      {
+        location: vscode.ProgressLocation.Notification,
+        title: 'Signing in to Deephaven...',
+        cancellable: true,
+      },
+      async (_, cancellationToken) => {
+        await vscode.env.openExternal(
+          vscode.Uri.parse(samlLoginUrl.toString())
```

Comment:
Should be using the `strict` option when parsing - it's "strongly advised": https://code.visualstudio.com/api/references/vscode-api#Uri.parse
```suggestion
          vscode.Uri.parse(samlLoginUrl.toString(), true)
```